shoebot/data/typography.py

```diff
@@ -178,7 +194,23 @@ def _pre_render(self):
         # self.layout.set_spacing(int(((self.lineheight-1)*self._fontsize)*Pango.SCALE)) #pango requires an int casting
         # we pass pango font description and the text to the pango layout
         self._pango_layout.set_font_description(self._pango_fontface)
-        self._pango_layout.set_text(self.text, -1)
+
+        if not self.markup_vars:
+            self._pango_layout.set_text(self.text, -1)
```

Comment:
This looks like an early bailout, which is good - its simplifies things.
Put a `return` here and ditch the `else` that follows and all the code below can lose an indent level.

Reply:
This `if` is also the boundry between two tests:  in theory at least, every `if` generates a new test.   Which might be why one reason test people try and avoid state where possible :)

The rest of this will be about avoiding state.

Where possible, state can be avoided by avoiding:
- `.append()` on `list`s, by using list comprehensions and filters.*
- avoiding `+=` to build strings using fstrings, and `.join` on strings.*
- avoiding setting elements in dicts*

* within reason !  these aren't hard and fast rules.

Reply:
I'll document attempting to refactor this so there's less code to test:

Maybe this can be refactored and use dict comprehensions / filters to reduce the manual dict building.
There are two things going on in this loop, so it makes sense to separate them:

1.  Copying named variables `("underline  ...`
2. Creating the `letter_spacing` variable from `tracking`

For 1:
It's a visit to the trusty search engine with the search terms:   `python dict filter certain items` and luckily there's this answer https://stackoverflow.com/questions/23862406/filter-items-in-a-python-dictionary-where-keys-contain-a-specific-string

In the end I couldn't see anything exactly right, so tried a couple of things in jupyter using the expected inputs + outputs and got it down to this:
```markup_vars = {arg: kwargs[arg] for arg in ["underline", "underlinecolor"] if arg in kwargs}```
(I was only testing with the first couple of arguments)

For 2:
To keep this simple it may as well adjust the dict, but is moved outside the loop:
```
markup_vars = {arg: kwargs[arg] for arg in ["underline", "underlinecolor"] if arg in kwargs}
if "tracking" in kwargs:
    markup_vars["letter_spacing"] = int(1024 * kwargs["tracking"])
```

(I haven't run this through Black, or added all the needed variables).